pilot/pkg/config/monitor/copilot_snapshot.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package monitor
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	copilotapi "code.cloudfoundry.org/copilot/api"
+
+	networking "istio.io/api/networking/v1alpha3"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/log"
+)
+
+// CopilotClient defines a local interface for interacting with Cloud Foundry Copilot
+//go:generate counterfeiter -o fakes/copilot_client.go --fake-name CopilotClient . CopilotClient
+type CopilotClient interface {
+	copilotapi.IstioCopilotClient
+}
+
+// CopilotSnapshot provides a snapshot of configuration from the Cloud Foundry Copilot component.
+type CopilotSnapshot struct {
+	store             model.ConfigStore
+	client            CopilotClient
+	hostnameBlacklist []string
+}
+
+// NewCopilotSnapshot returns a CopilotSnapshot used for integration with Cloud Foundry.
+// The store is required to discover any existing gateways: the generated config will reference those gateways
+// The client is used to discover Cloud Foundry Routes from Copilot
+// The snapshot will not return routes for any hostnames which end with strings on the hostnameBlacklist
+func NewCopilotSnapshot(store model.ConfigStore, client CopilotClient, hostnameBlacklist []string) *CopilotSnapshot {
+	return &CopilotSnapshot{
+		store:             store,
+		client:            client,
+		hostnameBlacklist: hostnameBlacklist,
+	}
+}
+
+// ReadConfigFiles returns a complete set of VirtualServices for all Cloud Foundry routes known to Copilot.
+// It may be used for the getSnapshotFunc when constructing a NewMonitor
+func (c *CopilotSnapshot) ReadConfigFiles() ([]*model.Config, error) {
+	resp, err := c.client.Routes(context.Background(), new(copilotapi.RoutesRequest))
```

Comment:
Does copilot have a default deadline on incoming requests? Is there a reasonable default we can set here? We shouldn't be making any calls w/o some kind of deadline set on the context as a matter of hygiene. Too easy to leak connections/goroutines on calls that never come back otherwise.

I think we haven't been doing a good job of checking for that kind of thing in the codebase to date, but we need to (especially as we're zeroing in on Pilot perf).

Reply:
@ZackButcher great catch - we'll just add a context with deadline here. Don't know what is reasonable for this call.

Reply:
We talked in Slack, but to capture discussion here for future reference: the specific value isn't really important so much as that there's _some_ deadline. A decent value is easy to get from running the system in prod for a bit (you're running copilot behind Envoy, to get those sweet, sweet Mixer metrics, right? :D ).